test/jdk/tools/launcher/Zip64ExecutableJarTest.java

```diff
@@ -0,0 +1,197 @@
+            forceCreateLargeZip64Entry(jaros);
+            // now that the large entry has been written as sparse holes, we now
+            // switch to writing the actual bytes for the rest of the entries
+            sos.sparse = false;
+            // add the main class
+            final JarEntry mainClassEntry = new JarEntry(MAIN_CLASS.replace('.', '/') + ".class");
+            jaros.putNextEntry(mainClassEntry);
+            jaros.write(Files.readAllBytes(mainClassFile));
+            jaros.closeEntry();
+            // finally add the META-INF/MANIFEST.MF
+            final Manifest manifest = new Manifest();
+            final Attributes mainAttributes = manifest.getMainAttributes();
+            mainAttributes.putValue("Manifest-Version", "1.0");
+            mainAttributes.putValue("Main-Class", MAIN_CLASS);
+            final JarEntry manifestEntry = new JarEntry("META-INF/MANIFEST.MF");
+            jaros.putNextEntry(manifestEntry);
+            manifest.write(jaros);
+            jaros.closeEntry();
+        }
+        return jarFile;
+    }
+
+    private static void forceCreateLargeZip64Entry(final JarOutputStream jaros) throws IOException {
+        final JarEntry entry = new JarEntry("entry-1");
+        entry.setMethod(STORED); // no need to deflate, we want the entry to be large sized
+        entry.setSize(ZIP64_SIZE_MAGICVAL);
+        // start with a dummy value, we will update it once the entry content is written
+        entry.setCrc(0);
+        jaros.putNextEntry(entry);
+        long numWritten = 0;
+        final byte[] entryContent = new byte[102400];
+        final CRC32 crc = new CRC32();
+        // keep writing the entry content till we reach the
+        // desired size that represents a zip64 entry
+        while (numWritten < ZIP64_SIZE_MAGICVAL) {
+            final long remaining = ZIP64_SIZE_MAGICVAL - numWritten;
+            final int len = remaining < entryContent.length
+                    ? (int) remaining
+                    : entryContent.length;
+            jaros.write(entryContent, 0, len);
+            numWritten += len;
+            crc.update(entryContent, 0, len);
+        }
+        entry.setCrc(crc.getValue()); // update the CRC in the entry
+        jaros.closeEntry();
+    }
+
+    /*
+     * An OutputStream which writes holes through its write() methods, until it is
+     * instructed to write the actual bytes. This implementation allows us to create large
+     * ZIP files without actually consuming large disk space.
+     * Instances of this class should be passed directly to the ZipOutputStream/JarOutputStream
+     * constructor, without any buffering, to allow for the implementation to correctly keep
+     * track of when to and when not to write holes.
+     */
+    private static class SparseOutputStream extends FilterOutputStream {
+        private final FileChannel channel;
+        private boolean sparse = true; // if true then contents will be written as sparse holes
```

Comment:
Would be nice if we could make the file produced in this test a valid ZIP file, that is narrowing the sparse mode scope to just the writing of the big entry data payload.

That way it's easier to inspect the file independently, using tools such as `zipdetails`

Consider fixing this by making `SparseOutputStream.sparse` false by default, setting it to true before the write loop and back to false after. That way the ZIP will include the Local File Header and the Data Descriptor for the entry.